### PR TITLE
Homework Cars 

### PR DESCRIPTION
**Pull request by kari0036**
- `accelerate()` only changes speed when `parameter >= 0` and `parameter < maxSpead_`

- added `speed_` variable to Car class and `maxSpead_`

- checks Gear, Engine

- The engine can no longer be replaced while driving. :/

- power, batteryCapacity, capacity, gears, currentGear, Engine are protected/private

- add tests

### CODE REVIEW

File: ElectricCar.cpp [hunk 1]
```diff
@@ -1,16 +1,48 @@
...
+    return electricEngine_->getPower();
 }
 
-ElectricCar::~ElectricCar()       { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::turnLeft()      { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::turnRight()     { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::brake()         { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::accelerate(int) { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::charge()        { std::cout << __FUNCTION__ << std::endl; }
+void ElectricCar::changeEngine(Engine* engine) {
```
Comment (WBigaj) [inline ElectricCar.cpp:35]: Can't you just take ElectricEngine as parameter?

**Review by WBigaj — COMMENTED** [review]
I have checked your code and have one question.

**Review by MaciejSalamonski — APPROVED** [review]
Nice code :). Have a good day!

File: ElectricCar-ut.cpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
+#include "catch2.hpp"
+
+#include "CarHaveMove.hpp"
+#include "ElectricCar.hpp"
+#include "PetrolEngine.hpp"
+#include "WrongEngine.hpp"
+
+SCENARIO("CorrectWorkElectricCar", "Cars"){
+    GIVEN("ElectricCar"){
+        ElectricCar tesla = new ElectricEngine(120, 700);
```
Comment (ziobron) [inline ElectricCar-ut.cpp:10]: implicit conversion. You should rather use:
```suggestion
        auto tesla = ElectricCar{new ElectricEngine(120, 700)};
```

File: ElectricCar.cpp (hunk 1 above)
Comment (ziobron) [reply ElectricCar.cpp:35]: In this case it wouldn't be a base class function overload.

File: ElectricCar.cpp [hunk 3]
```diff
@@ -1,16 +1,48 @@
...
+    if (speed_ == 0) {
+        if (typeid(*engine) != typeid(ElectricEngine)) {
+            throw InvalidEngine("Wrong type of Engine");
+            return;
```
Comment (ziobron) [inline ElectricCar.cpp:39]: return not needed when throwing

**Review by ziobron — APPROVED** [review]

**Comment by ziobron** [comment]
4 XP za wyjątek InvalidGear w Cars (3521)
5 * 2 XP za poprawkę interfejsu w Cars (3522)
10 XP za testy w Cars (3523)
3 * 2 XP za punktualność (1119)
{ "requirement_ids": [3521, 3523, 3522, 3522, 3522, 3522, 3522, 1119, 1119, 1119] }

**Pull request closed without merging**